src/drivers/cmfileapi-driver.ts

```diff
@@ -0,0 +1,218 @@
+
+  async doSetKit(need_clean: boolean, cb: () => Promise<void>): Promise<void> {
+    this._needsReconfigure = true;
+    if (need_clean) {
+      await this._cleanPriorConfiguration();
+    }
+    await cb();
+    if (!this.generator) {
+      throw new NoGeneratorError();
+    }
+  }
+
+  async asyncDispose() {
+    this._codeModelChanged.dispose();
+    this._cacheWatcher.dispose();
+  }
+
+  protected async doPreCleanConfigure(): Promise<void> { await this._cleanPriorConfiguration(); }
+
+  async doConfigure(args_: string[], outputConsumer?: proc.OutputConsumer): Promise<number> {
+    const api_path = this.getCMakeFileApiPath();
+    await createQueryFileForApi(api_path);
+
+    // Dup args so we can modify them
+    const args = Array.from(args_);
+    args.push('-H' + util.lightNormalizePath(this.sourceDir));
+    const bindir = util.lightNormalizePath(this.binaryDir);
+    args.push('-B' + bindir);
+    const gen = this.generator;
+    if (gen) {
+      args.push(`-G${gen.name}`);
+      if (gen.toolset) {
+        args.push(`-T${gen.toolset}`);
+      }
+      if (gen.platform) {
+        args.push(`-A${gen.platform}`);
+      }
+    }
+    const cmake = this.cmake.path;
+    log.debug('Invoking CMake', cmake, 'with arguments', JSON.stringify(args));
+    const env = await this.getConfigureEnvironment();
+    const res = await this.executeCommand(cmake, args, outputConsumer, {environment: env}).result;
+    log.trace(res.stderr);
+    log.trace(res.stdout);
+    if (res.retc == 0) {
+      this._needsReconfigure = false;
+    }
+    await this._reloadPostConfigure();
+    return res.retc === null ? -1 : res.retc;
+  }
+
+  async doPostBuild(): Promise<boolean> {
+    await this._reloadPostConfigure();
+    return true;
+  }
+
+  private getCMakeFileApiPath() { return path.join(this.binaryDir, '.cmake', 'api', 'v1'); }
+  private getCMakeReplyPath() {
+    const api_path = this.getCMakeFileApiPath();
+    return path.join(api_path, 'reply');
+  }
+
+  private async _reloadPostConfigure() {
+    const reply_path = this.getCMakeReplyPath();
+    const indexFile = await loadIndexFile(reply_path);
+    if (indexFile) {
+      this._generatorInformation = indexFile.cmake.generator;
+
+      // load cache
+      const cache_obj = indexFile.objects.find((value: index_api.Index.ObjectKind) => value.kind === 'cache');
+      if (!cache_obj) {
+        throw Error('No cache object found');
+      }
+
+      this._cache = await loadCacheContent(path.join(reply_path, cache_obj.jsonFile));
+
+      // load targets
+      const codemodel_obj = indexFile.objects.find((value: index_api.Index.ObjectKind) => value.kind === 'codemodel');
+      if (!codemodel_obj) {
+        throw Error('No code model object found');
+      }
+      this._target_map = await loadConfigurationTargetMap(reply_path, codemodel_obj.jsonFile);
+      this._codeModel = await loadExtCodeModelContent(reply_path, codemodel_obj.jsonFile);
```

Comment:
I think there would be value in allowing users to use the codemodel that is persisted to disk in the event that they don't want to reconfigure frequently.  (e.g. #984)  Is there any other way to load the code model today, or do we need to invent this concept in the driver?

Reply:
The idea of this PR was implementing the possibilities of the ServerAPI for FileAPI.
Let us restructure the driver to allow a simple CMake configuration check (expectation between CmakeTools (UI) and CMake driver). The FileAPI gives the possibility to load the old code model inside of the CMake driver. The problem is that we have no way to test if all Kit/Variant/Settings are as expected by the UI. The UI can only force the driver into a configuration.

I´m not sure if #984 and the possibility to use a preconfigured build folder by a new CMake-Driver instance should be part of this PR.
I do not see the required changes to make it simple possible.
Could we move this into a new issue?

Reply:
It's fine to move it to a new issue.  I just wanted to raise the question in case it was easy to do right now.

Reply:
I tested the new `doInit` behavior of the driver. I think it fixed issue #984, but there are so many dependencies.

Reply:
Just to highlight how bad #984 is: This issue may decide over moving from previous IDEs to vscode in our company or not. And I guess this is also the point for many other guys. Other features of vscode are superior, but since we have very large code base, reconfiguring and therefore performing clean builds all the time is seriously hurting the performance of our developers. Moreover I don't see a way to work around this than excluding the plugin completely from our workflow at the moment, but doing so carries other disadvantages with it.